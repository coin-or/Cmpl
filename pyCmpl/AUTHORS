Mike Steglich - Technical University of Applied Sciences Wildau, Germany - mike.steglich@th-wildau.de

	