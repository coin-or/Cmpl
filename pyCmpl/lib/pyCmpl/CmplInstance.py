#***********************************************************************
 #  This code is part of pyCMPL 
 #
 #  Copyright (C) 2013
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  pyCMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  pyCMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU Lesser General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  pyCMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU Lesser General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************

#!/usr/bin/python 

import os
import xml.dom.minidom as dom
import random
import cStringIO

from xml.sax.saxutils import unescape, escape


from CmplException import *


#*************** CmplInstance ***********************************
class CmplInstance(object):

	#*********** constructor **********
	def __init__(self):
		
		self.__cmplFile = ""
		self.__optionsList = []
		self.__dataString = ""
		self.__cmplDataList = {}
		self.__instStr = cStringIO.StringIO()
		self.__jobId = ""
		self.__cmplName = ""
		
	#*********** end constructor ******
	
	#*********** destructor ***********
	def __del__(self):
		self.__instStr.close()
	#*********** end destructor *******
		
	
	# getter **************************	
	@property
	def jobId(self):
		return self.__jobId
		
	@property
	def options(self):
		return self.__optionsList
	
	# end getter ***********************	
	
		
	#*********** cmplInstanceStr **********
	def cmplInstanceStr(self, cmplFileName, optList, dataString, jobId):
	
		if not os.path.isfile(cmplFileName):
			raise CmplException("CMPL file " + cmplFileName + " does not exist."  )
		
		self.__cmplFile = cmplFileName
		self.__optionsList = optList
		self.__jobId = jobId
		self.__dataString = dataString
	
		if self.__dataString!="":
			self.__cmplDataList.update({ "cmplData---"+os.path.basename(os.path.splitext(self.__cmplFile)[0])+".cdat" :   self.__dataString  })
			
		try:
			f = open(self.__cmplFile, "r")
			lines = f.readlines()
			f.close()

			commentSection=False
			
			lineNr = 0	
			for line in lines:
			
				line = line.strip()
			
				# besser contains
				if line.startswith('/*'):
					commentSection = True
					continue
					
				# besser contains	
				if line.startswith('*/'):
					commentSection = False
					
				if commentSection == True:
					continue
					
				if line.startswith('%data'):
					if line.find(':') != -1:
						tmpName = line[5:line.find(':')].strip()
					else:
						tmpName = line[5:].strip()
						
					if tmpName == "":	
						if self.__dataString!="":
							lines[lineNr] = line.replace("%data", "%data cmplData---"+os.path.basename(os.path.splitext(self.__cmplFile)[0])+".cdat ") + "\n"
							tmpName="cmplData---"
							
						else:				
							tmpName=os.path.basename(os.path.splitext(self.__cmplFile)[0])+".cdat" 
				
								
					if not (tmpName in self.__cmplDataList or tmpName=="cmplData---"):
					
						tmpName1=os.path.dirname(os.path.abspath(cmplFileName))+"/"+tmpName
						
						if not os.path.isfile(tmpName1):
							raise CmplException("cmplDataFile " + tmpName1 + " does not exist."  )
						
						f = open(tmpName1, "r")
						dataLines = f.readlines()
						f.close()		
					
						tmpString = cStringIO.StringIO()
						for dline in dataLines:
							tmpString.write(dline)
								
						self.__cmplDataList.update({ tmpName :  tmpString.getvalue()  })
						tmpString.close()
				
				lineNr += 1
				
					
			self.__instStr.write("<?xml version = \"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n")
			self.__instStr.write("<CmplInstance version=\"1.0\">\n")
			self.__instStr.write("<general>\n")
			self.__instStr.write("<name>"+os.path.basename(self.__cmplFile)+"</name>\n")
			self.__instStr.write("<jobId>"+self.__jobId+"</jobId>\n")
			self.__instStr.write("</general>\n")
			
			if len(self.__optionsList) > 0:
				self.__instStr.write("<options>\n")
				for opt in self.__optionsList:
					self.__instStr.write("<opt>"+opt+"</opt>\n")
				self.__instStr.write("</options>\n")
			
			self.__instStr.write("<problemFiles>\n")
			self.__instStr.write("<file name=\""+ escape(os.path.basename(self.__cmplFile))+ "\" type=\"cmplMain\">\n")
			
			
			for line in lines:
				self.__instStr.write(escape(line))
				#self.__instStr.write(b64encode(line))
			
			
			#self.__instStr.write("\n")	
			
			self.__instStr.write("</file>\n")	
			
			for d in self.__cmplDataList:
				self.__instStr.write("<file name=\""+ d + "\" type=\"cmplData\">\n")	
				#self.__instStr.write(b64encode(self.__cmplDataList[d]))
				self.__instStr.write(escape(self.__cmplDataList[d]))
				self.__instStr.write("\n")	
				self.__instStr.write("</file>\n")
			self.__instStr.write("</problemFiles>\n")	
			self.__instStr.write("</CmplInstance>\n")	
    		
		except IOError, e:
			raise CmplException("IO error : "+e)
		
		return self.__instStr.getvalue()
		
	#*********** end cmplInstanceStr ******	
		
	#*********** writeCmplInstance **********
	def writeCmplInstance(self, folder, instStr):	
	
		if os.path.exists(folder) == False:
			raise CmplException("Path <"+self.__cmplServerPath+"> doesn't exist.")
			
		instDom = dom.parseString(instStr)

		if instDom.firstChild.nodeName!="CmplInstance":
			raise CmplException("Cant't read cmplInstance file - wrong file type!")
			
		for entry in instDom.firstChild.childNodes: 
		
				if entry.nodeName == "general":
					for entry1 in entry.childNodes:
						if entry1.nodeName == "name":
							self.__cmplName = unescape(entry1.firstChild.data.strip())
							continue
						if entry1.nodeName == "jobId":
							self.__jobId = entry1.firstChild.data.strip()
							continue
							
				if entry.nodeName == "options":
					for entry1 in entry.childNodes: 
						if entry1.nodeName == "opt":
							self.__optionsList.append(entry1.firstChild.data.strip())
							
				if entry.nodeName == "problemFiles":
					for entry1 in entry.childNodes: 
						if entry1.nodeName == "file":
							tmpName = folder+self.__jobId+os.sep+entry1.getAttribute("name")
							#tmpContent = b64decode(entry1.firstChild.data.strip())
							tmpContent = unescape(entry1.firstChild.data.strip())
							
							try:
								f = open(tmpName, 'w')
								f.write(tmpContent)		
								f.close()
							except IOError, e:
								raise CmplException("IO error for file " + tmpName + ": "+e)
	#*********** end writeCmplInstance *******

#*************** end CmplInstance  ******************************

		