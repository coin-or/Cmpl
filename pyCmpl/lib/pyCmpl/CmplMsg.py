#***********************************************************************
 #  This code is part of pyCMPL 
 #
 #  Copyright (C) 2013
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  pyCMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  pyCMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU Lesser General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  pyCMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU Lesser General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************

#!/usr/bin/python 


import os
import re
import cStringIO 

from xml.sax.saxutils import unescape
from CmplDefs import *
from CmplException import *


#*************** CmplMsg *****************************************
class CmplMsg(object):
	def __init__(self):
		self.__type = ""
		self.__file = ""
		self.__line = ""
		self.__description = ""
		
	# getter and setter 	
	@property
	def type(self):
		return self.__type
	
	def setType(self, type):
		self.__type=type
		
	@property
	def file(self):
		return self.__file
		
	def setFile(self, file):
		self.__file=file
	
	@property
	def line(self):
		return self.__line
		
	def setLine(self, line):
		self.__line=line
		
	@property
	def description(self):
		return self.__description
		
	def setDescription(self, msg):
		self.__description=msg
	#end getter and setter
	
			
#*************** end CmplMsg *************************************


#*************** CmplMsg *****************************************
class CmplMessages(object):	

	def __init__(self, msgFile=None):
		self.__cmplStatus = ""
		self.__cmplVersion = ""
		self.__cmplMessage = ""
		self.__msgFile = msgFile
		self.__cmplMessageList = []
		
	#*********** cmplStatus ********	
	@property
	def cmplStatus(self):
		if self.__cmplStatus=="normal":
			return CMPL_OK
		elif self.__cmplStatus=="warning":
			return CMPL_WARNINGS
		elif self.__cmplStatus=="error":
			return CMPL_FAILED
		else:
			return CMPL_UNKNOWN
	#*********** end cmplStatus *****	
	
	#*********** cmplMessageList ****	
	@property
	def cmplMessageList(self):
		return self.__cmplMessageList
	#******* end cmplMessageList ****
	
	#*********** cmplMessage ********	
	@property
	def cmplMessage(self):
		return self.__cmplMessage
	#******* end cmplMessage ********
	
		
	#*********** readCmplMessages ********	
	def readCmplMessages(self, msgStr=None):
	
		if msgStr == None and self.__msgFile == None:
			raise CmplException("Neither cmplMessageFile nor cmplMessageString defined")
		
		if 	self.__msgFile != None:
			if not os.path.isfile(self.__msgFile):
				raise CmplException("CMPL failed."  )
	
			try:	
				f = open(self.__msgFile, "r")
				msgStr = f.read()
				f.close()
			except IOError, e:
				raise CmplServerException("IO error for instance file ")
				
				
		lines = cStringIO.StringIO(msgStr) 
		
		lineNr=1
		generalSection=False
		msgSection=False
		x = CmplMsg()
		
		for line in lines:
			if lineNr == 1:
				if not "<?xml version" in line: 
					raise CmplException("File " + self.__msgFile + " - is not a XML file!")
				else:
					lineNr += 1
					continue
			if lineNr == 2:
				if not "<CmplMessages" in line: 
					raise CmplException("Cant't read cmplMessage file " + self._cmplMsgFile + " - wrong file type!")
				else:
					lineNr += 1
					continue
	
			if "<general>" in line:
				generalSection=True
				msgSection=False
				continue
			if "</general>" in line:
				generalSection=False
				continue
			if "<messages" in line:
				generalSection=False
				msgSection=True
				continue
			if "</messages" in line:
				msgSection=False
				continue
				
			if generalSection:
				if "<generalStatus" in line:
					self.__cmplStatus = re.findall("<generalStatus>([^\"]*)</generalStatus>", line)[0]
					continue
				if "<message" in line:
					self.__cmplMessage = re.findall("<message>([^\"]*)</message>", line)[0]
					continue
				if "<cmplVersion" in line:
					self.__cmplVersion = re.findall("<cmplVersion>([^\"]*)</cmplVersion>", line)[0]
					continue
		
			if msgSection:
				if "<message" in line:
					tmpList=re.findall("\"([^\"]*)\"", line)
					x = CmplMsg()
					x.setType(tmpList[0])
					x.setFile(tmpList[1])
					x.setLine(tmpList[2])
					x.setDescription(unescape(tmpList[3]))
					self.__cmplMessageList.append(x)

						
							
	#*********** end readCmplMessages ****	
		
	

		
		
		
		

		