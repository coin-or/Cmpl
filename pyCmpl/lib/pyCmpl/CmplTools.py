#***********************************************************************
 #  This code is part of pyCMPL 
 #
 #  Copyright (C) 2013
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  pyCMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  pyCMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU Lesser General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  pyCMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU Lesser General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************

#!/usr/bin/python 

import re


#*************** CmplTools ***************************************
class CmplTools(object):

	#*********** stringContainsWhitespace **********
	@staticmethod
	def stringContainsWhitespace(string):
		regex = re.compile('[\t\n\x0b\x0c\r ]+')
		if regex.search(string):
			return True
		else:
			return False
	#*********** end stringContainsWhitespace ******
	
	#*********** strIsNumber ***********************
	@staticmethod
	def strIsNumber(str):
		r = re.compile('[^0-9.]+')
		if  r.search(str):
			return False
		else:
			return True
	#*********** end strIsNumber *******************
	
	#*********** strIsValidIndex ***********************
	@staticmethod
	def strIsValidIndex(str):
		r = re.compile('[^0-9a-zA-Z\-_.]+')
		if  r.search(str):
			return False
		else:
			return True
	#*********** end strIsValidIndex *******************
	
	
	
	#*********** strContains ***********************
	@staticmethod
	def strContains(str0, str1):
		return str1 in str0
	#*********** end strContains *******************

#*************** end CmplTools **********************************


		
		
		

		