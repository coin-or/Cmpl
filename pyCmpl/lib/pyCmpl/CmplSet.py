#***********************************************************************
 #  This code is part of pyCMPL 
 #
 #  Copyright (C) 2013
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  pyCMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  pyCMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU Lesser General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  pyCMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU Lesser General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************

#!/usr/bin/python 

from __future__ import division
from math import *

from CmplException import *
from CmplTools import *

	
#*************** CmplSet *****************************************		
class CmplSet(object):
	
	#*********** constructor ********
	def __init__(self, name, rank=1):
	
		self.__name = ""
		self.__type = 0   # 0 enumeration set - 1 tuple set - 2 alg set n..m - 3 alg set n(k)m
		self.__valueList = []
		self.__len = 0
		
		if type(name) != str:
			raise CmplException("not a valid name for set: " + str(name) )
		elif CmplTools.stringContainsWhitespace(name):
			raise CmplException("set name " + str(name) + " contains whitespaces")
		else:
			self.__name = name
			
		self.__rank=rank
	
	#*********** End constructor ****

	#*********** count **************	
	def count(self):
		_count = 0
		if self.__type == 0:  
			_count = len(self.__valueList)/self.__rank
		elif self.__type == 1:
			_count = len(self.__valueList)
		elif self.__type == 2:
			_count = self.__valueList[1]-self.__valueList[0]+1
		elif self.__type == 3:
			_count = ceil(  (fabs( self.__valueList[2] - self.__valueList[0])  +1 ) / fabs(self.__valueList[1]))
		return _count
	#*********** end count **********
		
			
	#*********** values *************	
	def setValues(self, val1, val2=None, val3=None):
		self.__valueList = []
	
		if val2 == None and val3 == None:
			if not 'LIST' in str(type(val1)).upper():
			#if type(val1) != list:
				raise CmplException("unexpected values for set " + self.__name + " : " + str(val1) + " is not a list")
			else:
				if 'LIST' in str(type(val1[0])).upper()  or 'TUPLE' in str(type(val1[0])).upper():
				#if type(val1[0]) == list or type(val1[0]) == tuple:
					if len(val1[0])!=self.__rank:
						raise CmplException("Rank and number of indexing entries for set " + self.__name + " : " + str(val1[0]) + " don't match.")
					self.__type = 1
					self.__len = len(val1) / self.__rank
				else:
					self.__type = 0
					if  self.__rank>1:
						raise CmplException( "incorrect definition of a tuple set " + self.__name)
				
				self.__valueList = val1
	
		elif val2 != None and val3 == None:
				self.__type = 2
				self.__rank = 1
				if type(val1) == int and type(val2) == int:
					if val2 < val1:
							raise CmplException( "unexpected values for set " + self.__name + " : end value "+ str(val2) + " is less then start value " + str(val1) )
					self.__valueList.append(val1)
					self.__valueList.append(val2)
				else:
					raise CmplException( "unexpected values for set " + self.__name + " : "+ str(val1) + "/" + str(val2) + "is not an valid integer combination")
	
		else:		
			self.__type = 3
			self.__rank = 1
			if type(val1) == int and type(val2) == int and type(val3) == int:
				self.__valueList.append(val1)
				self.__valueList.append(val2)	 
				self.__valueList.append(val3)

				if val2>0:
					if val3 <= val1:
						raise CmplException( "unexpected values for set " + self.__name + " : end value "+ str(val3) + " is less than or equal to start value " + str(val1) +" and the increment is positive "+ str(val2) )
				else: 
					if val2<0:
						if val1 <= val3:
							raise CmplException( "unexpected values for set " + self.__name + " : end value "+ str(val3) + " is greater than or equal to start value " + str(val1) +" and the increment is negative "+ str(val2) )
					else: 
						raise CmplException( "unexpected values for the set " + self.__name + " :  increment/decrement equals zero  " )
			else:
				raise CmplException( "unexpected values for set " + self.__name + " : "+ str(val1) + "/" + str(val2) + "/" + str(val3) + "is not an valid integer combination")
	#*********** end values *********
	
	
	#*********** valueList **********
	@property
	def valueList(self):
		return self.__valueList		
	#*********** End valueList ******	
	
	#*********** set **********
	@property
	def values(self):
	#	if self.__type == 0 or self.__type == 1:
	#		return self.__valueList		
		if self.__type == 0: 
			return self.__valueList		
			
		elif self.__type == 1:
			tmpList=[]
			for c in self.__valueList:
				tmpTuple="("
				
				for e in range(len(c)):
					if e>0:
						tmpTuple=tmpTuple+","
					if type(c[e])==str:	
						tmpTuple=tmpTuple+"\""+c[e]+"\""
					else:
						tmpTuple=tmpTuple+str(c[e])
				
				tmpTuple=tmpTuple+")"	
				tmpList.append(eval(tmpTuple))
			
			return tmpList
				
		elif self.__type == 2:
			return range(self.__valueList[0],self.__valueList[1]+1)
		elif self.__type == 3:
			if self.__valueList[1]>0:
				return range(self.__valueList[0],self.__valueList[2]+1,self.__valueList[1])
			else:
				return range(self.__valueList[0],self.__valueList[2]-1,self.__valueList[1])
	#*********** End valueList ******	
		
	#*********** name ***************
	@property
	def name(self):
		return self.__name		
	#*********** end name ***********	
	
	#*********** type ***************
	@property
	def type(self):
		return self.__type		
	#*********** end type ***********
	
	#*********** rank ***************
	@property
	def rank(self):
		return self.__rank		
	#*********** end rank ***********
	
	#*********** len ****************
	@property
	def len(self):
		if self.__type == 0 or self.__type == 1:
			return len(self.__valueList)	
		elif self.__type == 2:
			return len(range(self.__valueList[0],self.__valueList[1]+1))
		elif self.__type == 3:
			if self.__valueList[1]>0:
				return len(range(self.__valueList[0],self.__valueList[2]+1,self.__valueList[1]))
			else:
				return len(range(self.__valueList[0],self.__valueList[2]-1,self.__valueList[1]))
	#*********** end len ************
	
	
#*************** End CmplSet *****************************************		
	

	


		
		
		
		

		