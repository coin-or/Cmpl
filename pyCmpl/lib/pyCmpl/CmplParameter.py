#***********************************************************************
 #  This code is part of pyCMPL 
 #
 #  Copyright (C) 2013
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  pyCMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  pyCMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU Lesser General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  pyCMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU Lesser General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************

#!/usr/bin/python 

from __future__ import division
from math import *

from CmplException import *
from CmplTools import *
from CmplSet import *

		
#*************** CmplParameter ***********************************		
class CmplParameter(object):

	#*********** constructor *********
	def __init__(self, name, *sets ):
	
		self.__name = ""
		self.__rank = 0
		self.__valueList = []
		self.__count=1
		self.__defaultVal=None
		
		defSets=None
		
		if len(sets) == 0:
			defSets=sets
		else:
			if type(sets[0]) == list:
				defSets=sets[0]
			else:
				defSets=sets
	
		if type(name) != str:
			raise CmplException("not a valid name for a parameter: " + str(name) )
		elif CmplTools.stringContainsWhitespace(name):
			raise CmplException("parameter name " + str(name) + " contains whitespaces")
		else:
			self.__name = name
		
		self._setList = []
		self.__rank = 0
	
		for s in defSets:
			if type(s) != CmplSet:
				raise CmplException(str(s)+ " is not a CMPL set ")
			else: 
				self._setList.append(s)
				self.__rank+=s.rank
		
	#*********** End constructor ****
	
	#*********** valueCount *********	
	def __countElements(self, vList):
		_count=len(vList)
		if type(vList[0]) == list:
			_count *= self.__countElements(vList[0])
		return _count
	#*********** end valueCount *****
		
	
	#*********** values *************	
	def setValues(self, val , defaultVal=None):
		self.__valueList=[]
		if self.__rank == 0:
			if not 'LIST' in str(type(val)).upper():
			#if type(val) != list:
				self.__valueList.append(val)
			else:
				raise CmplException("incompatible data for a scalar parameter : " + str(val) )
		else:
			if  'LIST' in str(type(val)).upper():
			#if type(val) == list:
				self.__valueList=val
				sCount=1
				for s in self._setList:
					sCount*=s.count()
					
				self.__count = self.__countElements(val)		
				if self.__count !=  sCount :
					raise CmplException("The dimension of the paramter "+ self.__name +" doesn't match the dimension of the set(s)."  )

			elif  'DICT' in str(type(val)).upper():
			#elif type(val) == dict:
				self.__valueList=val
				
				# keys are to be checked here or not to be checked in here (then in CMPL)??

				if defaultVal!=None:
					#sparse array with default
					if not (type(defaultVal) != str or type(defaultVal) != int or type(defaultVal) != int or type(defaultVal) != float ):
						raise CmplException("Wrong type of the default value of the paramter "+ self.__name +" : " + str(defaultVal) + " : " + type(defaultVal)  )
						
					self.__defaultVal=defaultVal
				
				else:
					#dense array
					sCount=1
					sRank=0
					for s in self._setList:
						sCount*=s.count()
						
					if sCount!=len(val):
						raise CmplException("The dimension of the paramter "+ self.__name +" doesn't match the dimension of the set(s)."  )
					
					if 'LIST' in str(type(val.values()[0])).upper() or 'DICT' in str(type(val.values()[0])).upper():
					#if type(val.values()[0]) == list or type(val.values()[0])== dict:
						raise CmplException("Unexpected data type for paramter "+ self.__name   )
					
			else:
				raise CmplException("incompatible data for a parameter array: " + str(val) )
	
	#*********** end values *********
	
	#*********** setList ************	
	@property
	def setList(self) :
		return self._setList
	#*********** end setList ********
	
	#*********** valueList **********
	@property
	def values(self):
		return self.__valueList		
	#*********** End valueList ******	
		
	#*********** valueList **********
	@property
	def value(self):
		return self.__valueList[0]		
	#***	
		
	#*********** name ***************
	@property
	def name(self):
		return self.__name		
	#*********** end name ***********	
	
	#*********** rank ***************
	@property
	def rank(self):
		return self.__rank		
	#*********** end rank ***********
	
	#*********** len ***************
	@property
	def len(self):
		return self.__count		
	#*********** end rank ***********

	#*********** defaultValue ************	
	@property
	def defaultValue(self) :
		return self.__defaultVal
	#*********** end setList ********
		
#*************** end CmplParameter *******************************		
	
		
	

		
		
		
		

		