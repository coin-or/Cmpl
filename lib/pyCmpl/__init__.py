from CmplDefs import *
from Cmpl import *
from CmplException import *
from CmplMsg import *
from CmplTools import *
from CmplSolution import *
from CmplSet import *
from CmplParameter import *
from CmplInstance import *
from CmplInfo import *

if not sys.version_info[:2] == (2,7):
	print "ERROR - pyCmpl only supports Python 2.7"
	exit(-1)
	