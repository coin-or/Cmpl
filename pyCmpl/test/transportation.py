#***********************************************************************
 #  This code is part of CMPL
 #
 #  Copyright (C) 2007, 2008, 2009, 2010, 2011, 2012, 2013
 #  Thomas Schleiff - Halle(Saale), Germany and
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  CMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  CMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  CMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************

#!/usr/bin/python 

from pyCmpl import *

try: 
	model = Cmpl("transportation.cmpl")
	
	routes = CmplSet("routes",2)
	routes.setValues([[1,1],[1,2],[1,4],[2,2],[2,3],[2,4],[3,1],[3,3]])

	plants = CmplSet("plants")
	plants.setValues(1,3)
	
	centers = CmplSet("centers")
	centers.setValues(1,4)
	
	costs = CmplParameter("c",routes)
	costs.setValues([3,2,6,5,2,3,2,4])
	
	s = CmplParameter("s",plants)
	s.setValues([5000,6000,2500])
	
	d = CmplParameter("d",centers)
	d.setValues([6000,4000,2000,2500])
	
	model.setSets(routes, plants, centers)
	model.setParameters(costs,s,d)
		
	model.setOption("%display nonZeros")
	model.solve() 
			
	if model.solverStatus == SOLVER_OK:
		model.solutionReport()
	else:
		print "Solver failed " + model.solver + " " + model.solverMessage
	
except CmplException, e:
	print e.msg
	
	
