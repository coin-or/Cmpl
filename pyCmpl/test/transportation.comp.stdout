---------------------------------------------------------------------------------------------------------
Problem              transportation.cmpl
Nr. of variables     8
Nr. of constraints   7
Objective name       costs
Solver name          CBC
Display variables    nonZeroVariables (all)
Display constraints  nonZeroConstraints (all)
---------------------------------------------------------------------------------------------------------

Objective status     optimal
Objective value      36500.00            (min!)

Variables           
Name                 Type            Activity          LowerBound          UpperBound            Marginal
---------------------------------------------------------------------------------------------------------
x[1,1]                  C             2500.00                0.00                 inf                0.00
x[1,2]                  C             2500.00                0.00                 inf                0.00
x[2,2]                  C             1500.00                0.00                 inf                0.00
x[2,3]                  C             2000.00                0.00                 inf                0.00
x[2,4]                  C             2500.00                0.00                 inf                0.00
x[3,1]                  C             2500.00                0.00                 inf                0.00
---------------------------------------------------------------------------------------------------------

Constraints         
Name                 Type            Activity          LowerBound          UpperBound            Marginal
---------------------------------------------------------------------------------------------------------
supplies[1]             E             5000.00             5000.00             5000.00                3.00
supplies[2]             E             6000.00             6000.00             6000.00                6.00
supplies[3]             E             2500.00             2500.00             2500.00                2.00
demands[1]              L             5000.00                -inf             6000.00                0.00
demands[2]              L             4000.00                -inf             4000.00               -1.00
demands[3]              L             2000.00                -inf             2000.00               -4.00
demands[4]              L             2500.00                -inf             2500.00               -3.00
---------------------------------------------------------------------------------------------------------

